### PR TITLE
Task/user snippets iteration 0

### PR DESCRIPTION
**Pull request by yonip23**
Adding an option to save a code snippet - Currently available for only internally in Tabnine's team:
image

### CODE REVIEW

File: src/commandsHandler.ts [hunk 1]
```diff
@@ -8,9 +8,11 @@ import {
 import { configuration } from "./binary/requests/requests";
 import setState from "./binary/requests/setState";
 import { Capability, isCapabilityEnabled } from "./capabilities/capabilities";
+import handleSaveSnippet from "./saveSnippetHandler";
 
 export const CONFIG_COMMAND = "TabNine::config";
 export const STATUS_BAR_COMMAND = "TabNine.statusBar";
+export const SAVE_SNIPPET_COMMAND = "Tabnine.saveSnippet";
```
Comment (boaz-codota) [inline src/commandsHandler.ts:15]: Maybe you could align all tasks to have an identical "format"? Whichever you choose.

File: src/binary/requests/saveSnippet.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  code: string;
+  filename: string;
+  start_offset: number;
+  end_offset: number;
+};
+
+export type SuccessSaveSnippetResponse = "Success";
+
+export type ErrorSaveSnippetResponse = {
+  Error: string;
```
Comment (boaz-codota) [inline src/binary/requests/saveSnippet.ts:11]: Why capital `Error`?

File: src/binary/requests/saveSnippet.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+export type SaveSnippetRequest = {
+  code: string;
+  filename: string;
+  start_offset: number;
+  end_offset: number;
+};
+
+export type SuccessSaveSnippetResponse = "Success";
```
Comment (boaz-codota) [inline src/binary/requests/saveSnippet.ts:8]: Results should always be objects, for them to be future proof.

File: package.json [hunk 4]
```diff
@@ -539,4 +549,4 @@
   "lint-staged": {
     "*.{ts,js,css,md}": "prettier --write src/"
   }
-}
+}
```
Comment (boaz-codota) [inline package.json:552]: Is this a linux thingy?

File: src/test/suite/saveSnippet.test.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    mockSaveSnippetResponse(aSaveSnippetSuccessResponse());
+
+    await vscode.commands.executeCommand(`${SAVE_SNIPPET_COMMAND}`);
+
+    assertNotification((stub) =>
+      stub.calledWithExactly(SUCCESS_MESSAGE, OK_BUTTON)
+    );
+  });
+
+  test("Shows an error notification when saving snippet fails", async () => {
```
Comment (boaz-codota) [inline src/test/suite/saveSnippet.test.ts:36]: Those tests are looking fine AF 👏

File: src/saveSnippetHandler.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    start_offset: document.offsetAt(selection.start),
+    end_offset: document.offsetAt(selection.end),
+  };
+
+  const result = await saveSnippet(request);
+
+  const error = result
+    ? (result as ErrorSaveSnippetResponse).Error
+    : NO_RESPONSE_ERROR_MESSAGE;
+  const message = error ? `${ERROR_MESSAGE_PREFIX}: ${error}` : SUCCESS_MESSAGE;
```
Comment (boaz-codota) [inline src/saveSnippetHandler.ts:28]: This code is ugly and impossible to read. Refactor it.

**Review by boaz-codota — APPROVED** [review]
Overall looks good

File: src/binary/requests/saveSnippet.ts (hunk 2 above)
Comment (yonip23) [reply src/binary/requests/saveSnippet.ts:11]: This is how the json is being received from the binary (enum variant)

File: src/binary/requests/saveSnippet.ts (hunk 3 above)
Comment (yonip23) [reply src/binary/requests/saveSnippet.ts:8]: how do i do that? the json response is `"Success"`

File: package.json (hunk 4 above)
Comment (yonip23) [reply package.json:552]: idk

File: src/saveSnippetHandler.ts (hunk 6 above)
Comment (yonip23) [reply src/saveSnippetHandler.ts:28]: lol sababa

**Comment by boaz-codota** [comment]
Also, I'd love to see our unicode icon as part of the action text in the dropdown menu (same as in the suggestions we provide)

File: src/saveSnippetHandler.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+export async function enableSaveSnippetContext(): Promise<void> {
+  await commands.executeCommand(
+    "setContext",
+    "tabnine.save-snippets:enabled",
+    true
+  );
+}
+
+function getErrorMessage(
```
Comment (dimacodota) [inline src/saveSnippetHandler.ts:44]: `getBInaryMessage`?

Comment (yonip23) [reply src/saveSnippetHandler.ts:44]: i think `getErrorMessage` is a more correct name 🤷🏻‍♂️

**Pull request merged**